valhalla/userrequests/state_changes.py

```diff
@@ -141,6 +145,93 @@ def modify_ipp_time_from_requests(ipp_val, requests_list, modification='debit'):
+    active_blocks = False
+    future_blocks = False
+    now = timezone.now()
+    for block in request_blocks:
+        start_time = dateutil.parser.parse(block['start']).replace(tzinfo=timezone.utc)
+        end_time = dateutil.parser.parse(block['end']).replace(tzinfo=timezone.utc)
+        if all([molecule['complete'] for molecule in block['molecules']]):
+            return 'COMPLETED'
+        if (not block['canceled'] and not any([molecule['failed'] for molecule in block['molecules']])
+            and start_time < now < end_time):
+            active_blocks = True
+        if now < start_time:
+            future_blocks = True
+
+    if not (future_blocks or active_blocks):
+        return 'FAILED'
+
+    return request_state
+
+
+def get_request_state(request_state, request_blocks, ur_expired):
+    if request_state == 'COMPLETED':
+        return request_state
+    else:
+        pond_state = get_request_state_from_pond_blocks(request_state, request_blocks)
+        if pond_state in ['COMPLETED', 'WINDOW_EXPIRED', 'CANCELED']:
+            return pond_state
+
+        if ur_expired:
+            return 'WINDOW_EXPIRED'
+
+        if pond_state == 'FAILED' and request_state in ['WINDOW_EXPIRED', 'CANCELED']:
+            return request_state
+
+        return pond_state
+
+
+def aggregate_request_states(request_states, operator):
+    # Failed states are equivalent to PENDING from UR perspective
+    request_states = [rs.replace('FAILED', 'PENDING') for rs in request_states]
+
+    # Set the priority ordering - assume AND by default
+    state_priority = ['WINDOW_EXPIRED', 'PENDING', 'COMPLETED', 'CANCELED']
+    if operator == 'ONEOF':
+        state_priority = ['COMPLETED', 'PENDING', 'WINDOW_EXPIRED', 'CANCELED']
+    elif operator == 'MANY':
+        state_priority = ['PENDING', 'COMPLETED', 'WINDOW_EXPIRED', 'CANCELED']
+
+    for state in state_priority:
+        if state in request_states:
+            return state
+
+    raise AggregateStateException('Unable to Aggregate States: {}'.format(request_states))
+
+
+def update_request_states_from_pond_blocks(pond_blocks):
+    sorted_blocks = sorted(pond_blocks, key=lambda x: x['molecules'][0]['tracking_number'])
+    blocks_by_tracking_num = itertools.groupby(sorted_blocks, lambda x: x['molecules'][0]['tracking_number'])
+    now = timezone.now()
+
+    for tracking_num, blocks in blocks_by_tracking_num:
+        sorted_blocks_by_request = sorted(blocks, key=lambda x: x['molecules'][0]['request_number'])
+        blocks_by_request_num = {k: list(v) for k,v in itertools.groupby(sorted_blocks_by_request, key=lambda x: x['molecules'][0]['request_number'])}
+        user_request = UserRequest.objects.get(pk=tracking_num)
+        ur_expired = user_request.max_window_time < now
+        request_states = []
+        requests = user_request.requests.all()
+        for request in requests:
+            new_r_state = request.state
```

Comment:
This took me a bit to understand. I think we should just add an `else` block that does `request_states.append(request.state)` this makes it a bit easier to realize that you're just adding the unchanged state.